src/core_functions/scalar/list/list_reduce.cpp

```diff
@@ -0,0 +1,241 @@
+#include "duckdb/core_functions/scalar/list_functions.hpp"
+#include "duckdb/core_functions/lambda_functions.hpp"
+#include "duckdb/planner/expression/bound_cast_expression.hpp"
+#include "duckdb/planner/expression/bound_function_expression.hpp"
+
+namespace duckdb {
+
+struct ReduceExecuteInfo {
+	ReduceExecuteInfo(LambdaFunctions::LambdaInfo &info, ClientContext &context) : left_slice(*info.child_vector) {
+		SelectionVector left_vector(info.row_count);
+		active_rows.Resize(0, info.row_count);
+		active_rows.SetAllValid(info.row_count);
+
+		right_sel.Initialize(info.row_count);
+		left_sel.Initialize(info.row_count);
+		active_rows_sel.Initialize(info.row_count);
+
+		idx_t old_count = 0;
+		idx_t new_count = 0;
+
+		// Initialize the left vector from list entries and the active rows
+		auto data = active_rows.GetData();
+
+		for (idx_t entry_idx = 0; old_count < info.row_count; entry_idx++) {
+			if (data[entry_idx] == 0) {
+				old_count += 64;
+				continue;
+			}
+
+			for (idx_t j = 0; j < 64 && entry_idx + j < info.row_count; j++) {
+				auto list_column_format_index = info.list_column_format.sel->get_index(old_count);
+				if (info.list_column_format.validity.RowIsValid(list_column_format_index)) {
+					if (info.list_entries[list_column_format_index].length == 0) {
+						throw ParameterNotAllowedException("Cannot perform list_reduce on an empty input list");
+					}
+					left_vector.set_index(new_count, info.list_entries[list_column_format_index].offset);
+					new_count++;
+				} else {
+					// Remove the invalid rows
+					info.result_validity->SetInvalid(old_count);
+					active_rows.SetInvalid(old_count);
+				}
+				old_count++;
+			}
+		}
+
+		left_slice.Slice(left_vector, new_count);
+
+		if (info.has_index) {
+			input_types.push_back(LogicalType::BIGINT);
+		}
+		input_types.push_back(left_slice.GetType());
+		input_types.push_back(left_slice.GetType());
+		for (auto &entry : info.column_infos) {
+			input_types.push_back(entry.vector.get().GetType());
+		}
+
+		expr_executor = make_uniq<ExpressionExecutor>(context, *info.lambda_expr);
+	};
+	ValidityMask active_rows;
+	Vector left_slice;
+	unique_ptr<ExpressionExecutor> expr_executor;
+	vector<LogicalType> input_types;
+
+	SelectionVector right_sel;
+	SelectionVector left_sel;
+	SelectionVector active_rows_sel;
+};
+
+static bool ExecuteReduce(idx_t loops, ReduceExecuteInfo &execute_info, LambdaFunctions::LambdaInfo &info,
+                          DataChunk &result_chunk) {
+	idx_t old_count = 0;
+	idx_t new_count = 0;
+	idx_t valid_count = 0;
+
+	// create selection vectors for the left and right slice
+	auto data = execute_info.active_rows.GetData();
+
+	for (idx_t entry_idx = 0; old_count < info.row_count; entry_idx++) {
+		if (data[entry_idx] == 0) {
+			old_count += 64;
+			continue;
+		}
+
+		for (idx_t j = 0; j < 64 && entry_idx + j < info.row_count; j++) {
```

Comment:
This condition looks off: `entry_idx + j < info.row_count`. Shouldn't it be `entry_idx * bits_per_entry + j < info.row_count`? Can we get a test for this? I think we need a list with more than 64 elements, and we need to verify the result.

Nit: `sizeof(idx_t) * 8` or `bits_per_entry` instead of `64`

Reply:
We do already have a test for very large lists:

```sql
statement ok
CREATE TABLE large_lists AS SELECT range % 4 g, list(range) l FROM range(10000) GROUP BY range % 4;

query I
SELECT list_reduce(l, (x, y) -> least(x, y)) FROM large_lists ORDER BY g;
----
0
1
2
3

query I
SELECT list_reduce(l, (x, y) -> x + y) FROM large_lists ORDER BY g;
----
12495000
12497500
12500000
12502500
```